content/blog/2019-10-30-loop-reduction/index.md

```diff
@@ -0,0 +1,375 @@
+
+The procedure of performing strength reduction is as follows:
+
+1. Create new variable: `A'`
+2. Initialize in preheader: `A’ = c1 * B + c2`
+3. Track value of B: add after `B=B+x`: `A’=A’+x*c1`
+4. Replace assignment to A: replace lone `A=...` with `A=A’`
+
+Thus, the key idea here is to first find out each induction variable `A` and then replace definition of A when executed.
+
+To find out each induction variable, we scan through the code to
+
+1. find out all the basic induction variables `B`
+2. find out all induction variables `A` in family of `B`, where `A` refers to the `B` at the time of definition
+
+The `A` here should be in one of the following conditions:
+i. `A` has a single assignment in the loop `L` in the form of:
+​    `A = B * c` | `A = c * B` | `A = B / c` | `A = B + c` | `A = c + B` | `A = B - c` | `A = c - B`
+ii. `A` has a single assignment in the loop `L` in the form of (`D` is an induction variable in the family of `B`) 
+​    `A = D * c` | `A = c * D` | `A = D / c` | `A = D + c` | `A = c + D` | `A = D - c` | `A = c - D`
+Also, no definitions of `D` outside `L` reaches the assignment of `A`, and every path between the point of assignment to `D` in `L` and the assignment to `A` has the same sequence (possibly empty) of definitions of `B`.
+
+After all induction variables are found, strength reduction is performed to add new initialization to the variable and reduce multiplications to additions following the procedure we have described above.
+
+
+
+# Challenges
+
+1. There are more properties we need than we originally expected. At first, we only generated loops, reaching variables. Then we for loop invariants code motion, we needed exits to the loop, dominance relations, live variables. 
+2. The representation of different variables are randomly decided at first and need implementation after we finalized the representation. For example,the loop invariant at first was stored as the list of instructions. Later, we found it necessary to change the storage form to the dictionary whose key is the block name. Otherwise, we would need to search and match the instruction to block, e.g, in the `move_LI` function.
+3. There are more similarity between basic induction variables and their families. It is sometimes tricky to differentiate them. Thus, the definition flow of each induction variable is maintained to tell them apart.
+
+
+
+## Evaluation
+
+1. We would like to evaluate our optimizer in two aspects: theoretical improvement and actual wall-clock speedup.
+
+   ### Benchmarks
+
+   Because loop invariant code motion and strength reduction are specific optimization for loops, and studies are based on case by case, we can hardly generate automatic testing scripts while writing bril codes to generate `for` loop are actually painful. Instead we reply on Typescript for help. 
```

Comment:
- "studies are based on case by case": I'm not entirely sure what this means?
- capitalize "Bril"
- "writing Bril codes to generate `for` loop": But Bril code doesn't *generate* `for` loops… rather, `for` loops in higher-level languages generate Bril code!
- reply -> rely
- capitalize "TypeScript"